### PR TITLE
[EXPERIMENT][backport] ObjectTpeJava

### LINKED ISSUE #12776: no type parameters for method handle: (x$1: java.util.function.BiFunction[_ >: T, Throwable, _ <: U])
## Reproduction steps

Scala version: 2.12.17

Works in 2.13

https://github.com/akka/akka/pull/31918/files#diff-f9740bf92e4d83e90e297bde9ecbbb5a624b269455cc21a0c805fd7a0458785cR71-R72

```scala
  def afterCompletionStage[T](duration: FiniteDuration, using: Scheduler)(value: => CompletionStage[T])(
      implicit ec: ExecutionContext): CompletionStage[T] =
    if (duration.isFinite && duration.length < 1) {
      try value
      catch { case NonFatal(t) => Futures.failedCompletionStage(t) }
    } else {
      val p = new CompletableFuture[T]
      using.scheduleOnce(duration) {
        try {
          val future = value
          future.handle(new BiFunction[T, Throwable, Unit] {
            override def apply(t: T, ex: Throwable): Unit = {
              if (t != null) p.complete(t)
              if (ex != null) p.completeExceptionally(ex)
            }
          })
        } catch {
          case NonFatal(ex) => p.completeExceptionally(ex)
        }
      }
      p
    }
```

## Problem

```
[error] C:\Users\hepin\IdeaProjects\akka\akka-actor\src\main\scala\akka\pattern\FutureTimeoutSupport.scala:71:18: no type parameters for method handle: (x$1: java.util.function.BiFunction[_ >: T, Throwable, _ <: U])java.util.concurrent.CompletionStage[U] exist so that it can be applied to arguments (java.util.function.BiFunction[T,Throwable,Unit])
[error]  --- because ---
[error] argument expression's type is not compatible with formal parameter type;
[error]  found   : java.util.function.BiFunction[T,Throwable,Unit]
[error]  required: java.util.function.BiFunction[_ >: T, Throwable, _ <: ?U]
[error] Note: T <: Any, but Java-defined trait BiFunction is invariant in type T.
[error] You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
[error]           future.handle(new BiFunction[T, Throwable, Unit] {
[error]                  ^
[error] C:\Users\hepin\IdeaProjects\akka\akka-actor\src\main\scala\akka\pattern\FutureTimeoutSupport.scala:71:25: type mismatch;
[error]  found   : java.util.function.BiFunction[T,Throwable,Unit]
[error]  required: java.util.function.BiFunction[_ >: T, Throwable, _ <: U]
[error]           future.handle(new BiFunction[T, Throwable, Unit] {
[error]                         ^
[error] two errors found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 16 s, completed 2023-4-20 23:43:53
[IJ]
```

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12776

### CODE REVIEW

**Comment by He-Pin** [comment]
Wow! Thank you very much. I was finding which PR is related .

File: test/files/pos/t12776.scala [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+import java.util.concurrent.CompletableFuture
+
+class C {
+  def f = new CompletableFuture[String].handle((s, t) => ())
+}
```
Comment (He-Pin) [inline test/files/pos/t12776.scala:6]: Yes it is!

**Comment by som-snytt** [comment]
This is educational and hopefully helpful to those stuck on 2.12, but might take some effort for polish and might also prove too much to ask in review. I'm not standing on the shoulders of giants but hanging onto their coattails.

The source commits and discussion were helpful. I learned that "I'll just tweak this thing" makes backporting more difficult.

It would be nice if all versions had a unified test corpus. Maybe it would be difficult to see which tests applied per version, as is true for JVM-dependent tests.

**Comment by som-snytt** [comment]
Squashed, since it was ignored anyway. Maybe they will cut a 2.12.19.

**Comment by SethTisue** [comment]
We'll take a look after 2.12.18 ships.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -3186,6 +3195,7 @@ trait Types
        */
       def unifySimple = {
         val sym = tp.typeSymbol
+        // TODO isAnyExactly
```
Comment (som-snytt) [inline src/reflect/scala/reflect/internal/Types.scala:3198]: TODO follow up what the TODO means

**Comment by som-snytt** [comment]
When the time comes, let me know if it's worth the trouble. Is there a client who will pay for review time?  This feels like much longer than a month ago, and it's doable but not trivial.

**Comment by lrytz** [comment]
I do think this is too risky to touch in 2.12. The workaround is the type parameter `new CompletableFuture[String].handle[Unit]((s, t) => ())`, right?

**Comment by som-snytt** [comment]
I'll leave it open for a bit, because the calculus might be: At this point, there is no reason to upgrade an old project on 2.12; a risk-averse project is still on some previous 2.12.x; only projects experiencing the pain point will upgrade and assume the costs of taking on the risk (which is always non-zero but perhaps the upgrade is assumed to be anodyne).

**Comment by SethTisue** [comment]
Well, as often with backports, this is a judgment call one could argue either way. With some reluctance, as I don't like to waste work that Som or anyone puts in, I think it's better not to gamble on this, especially since it's straightforward to work around.

**Comment by som-snytt** [comment]
If lives were at risk, anyone needing this can trivially build with it. Also they can hire me to support it.

**Comment by He-Pin** [comment]
Thank you very much @som-snytt

**Pull request closed without merging**